drivers/serial/uart_sy1xx.c

```diff
@@ -0,0 +1,317 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright (c) 2024 sensry.io
+ */
+
+#define DT_DRV_COMPAT sensry_sy1xx_uart
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/uart.h>
+#include <soc.h>
+#include <zephyr/sys/printk.h>
+#include <udma.h>
+#include <pad_ctrl.h>
+
+struct sy1xx_uart_config {
+	uint32_t base;
+	uint32_t inst;
+};
+
+typedef struct {
+	uint16_t data_len;
+	uint8_t *data;
+} uartTransfer_t;
+
+typedef enum {
+	DRIVERS_UART_STOP_1,
+	DRIVERS_UART_STOP_1_5,
+	DRIVERS_UART_STOP_2
+} uart_stop_t;
+
+typedef enum {
+	DRIVERS_UART_PAR_NONE,
+	DRIVERS_UART_PAR_EVEN,
+	DRIVERS_UART_PAR_ODD,
+	DRIVERS_UART_PAR_MARK,
+	DRIVERS_UART_PAR_SPACE
+} uart_parity_t;
+
+typedef struct {
+	uint32_t baudrate;
+	uart_stop_t stopbits;
+	uart_parity_t parity;
+} uartConfig_t;
+
+#define DEVICE_MAX_BUFFER_SIZE (512)
+
+struct sy1xx_uart_data {
+	uint8_t write[DEVICE_MAX_BUFFER_SIZE];
+	uint8_t read[DEVICE_MAX_BUFFER_SIZE];
+};
+
+/* prototypes */
+static int32_t drivers_uart_read(const struct device *dev, uartTransfer_t *request);
+static int32_t drivers_uart_write(const struct device *dev, uartTransfer_t *request);
+
+static int32_t drivers_uart_configure(const struct device *dev, uartConfig_t *uart_cfg)
+{
+	struct sy1xx_uart_config *config = (struct sy1xx_uart_config *)dev->config;
+
+	if (uart_cfg->baudrate == 0) {
+		return -1;
+	}
+
+	/*
+	 * The counter in the UDMA will count from 0 to div included
+	 * and then will restart from 0, so we must give div - 1 as
+	 * divider
+	 */
+	int32_t divider = soc_get_peripheral_clock() / uart_cfg->baudrate - 1;
+
+	/*
+	 * [31:16]: clock divider (from SoC clock)
+	 * [9]: RX enable
+	 * [8]: TX enable
+	 * [3]: stop bits  0 = 1 stop bit
+	 *                 1 = 2 stop bits
+	 * [2:1]: bits     00 = 5 bits
+	 *                 01 = 6 bits
+	 *                 10 = 7 bits
+	 *                 11 = 8 bits
+	 * [0]: parity
+	 */
+
+	/* default: both tx and rx enabled; 8N1 configuration; 1 stop bits */
+	volatile uint32_t setup = 0x0306 | uart_cfg->parity;
+
+	setup |= ((divider) << 16);
+	UDMA_WRITE_REG(config->base, UDMA_SETUP_REG, setup);
+
+	/* start initial reading request to get the dma running */
+	uint8_t dummy_data[10];
+
+	uartTransfer_t dummy_request = {
+		.data_len = 10,
+		.data = (uint8_t *)dummy_data,
+	};
+
+	drivers_uart_read(dev, &dummy_request);
+	return 0;
+}
+
+/*
+ * \return: < 0 ... error
+ *	    = 0 ok
+ *	    > 0 ... busy
+ */
```

Comment:
I recommend changing this into proper doxygen style comments so IDE can display it during code completion, especially the return values.

ditto for those below.

Reply:
done